internal/worker/run.go

```diff
@@ -191,7 +211,89 @@ func (w *Worker) runStep(ctx context.Context, log logr.Logger, sandbox cri.CRI,
 	log.Info(aurora.Green("Step completed").String())
 }
 
-func (w *Worker) runTaskLogic(ctx context.Context, syslog, log logr.Logger, slot int, claim *tcqueue.TaskClaim, wr io.Writer) error {
+func (w *Worker) uploadArtifact(
+	ctx context.Context, log logr.Logger, claim *tcqueue.TaskClaim,
+	rootContainer cri.Container, artifact *config.Artifact,
+	deps []<-chan error, ch chan<- error,
+) {
+	var err error
+	defer func() {
+		ch <- err
+		close(ch)
+	}()
+
+	// Wait for all dependencies.
+	for _, ch := range deps {
+		err = <-ch
+		if err != nil {
+			return
+		}
+	}
+
+	if artifact.Type != config.FileArtifact {
+		err = fmt.Errorf("only file artifacts are supported")
+		return
+	}
+
+	srcPath := artifact.Path
+	if !path.IsAbs(srcPath) {
+		srcPath = path.Join("/workspace", artifact.Path)
+	}
+
+	var r io.ReadCloser
+	r, info, err := rootContainer.ReadFiles(ctx, srcPath)
+	if err != nil {
+		return
+	}
+	defer func() {
+		if err := r.Close(); err != nil {
+			log.Error(err, "failed to close artifact file")
+		}
+	}()
+
+	if info.IsDir() {
+		err = fmt.Errorf("directory artifacts are not supported")
+		return
+	}
+
+	expiresAt := time.Time(claim.Task.Expires)
+	size := info.Size()
+
+	tr := tar.NewReader(r)
```

Comment:
is this reading the artifact as a tar file? does that mean all artifacts must be tar files?

Reply:
This is actually an unfortunate implementation detail of Docker's `ReadFiles`. This probably sits the wrong side of the interface but I'd prefer to fix that later.

Reply:
Ahh. It would be _nice_ if this _was_ on the other side... but fair enough. That isn't a now problem I don't think.